### PR TITLE
FHIR Package Loader Rewrite

### PR DESCRIPTION
**Pull request by cmoesel**
**Description:** _And now for something completely different..._

This PR represents a total rewrite of the FHIR Package Loader -- with completely new interfaces, implementations, and APIs. This rewrite was necessary to address limitations in the package loader as well as support future use cases (such as transitive dependencies, multiple versions of the the same package, etc).

The new FHIR Package Loader now utilizes an in-memory database (currently [sql.js](https://www.npmjs.com/package/sql.js)) to store and query metadata about packages and their resources. This reduces memory usage per-resource since now only the metadata is kept in memory (vs the entire JSON object). With more efficient memory usage, we can now load all resources from the package (instead of just conformance resources) and we will also be able to load transitive packages in the future.

Separate concerns of the package loader have been separated out into their own interfaces and implementations, including:
* `PackageDB` interface w/ `SQLJSPackageDB` implementation. Note that sql.js was chosen due to its support for synchronous queries -- which causes the least disruption to SUSHI (which has many synchronous functions).
* `RegistryClient` interface w/ `FHIRRegistryClient`, `NPMRegistryClient`, `RedundantRegistryClient`, and `DefaultRegistryClient` implementations
* `CurrentBuildClient` interface w/ `BuildDotFhirDotOrgClient` implementation
* `PackageCache` interface w/ `DiskBasedPackageCache` implementation
* `VirtualPackage` interface w/ `DiskBasedVirtualPackage` and `InMemoryVirtualPackage` implementations
* `PackageLoader` interface w/ `BasePackageLoader` implementation that accepts implementations of the previous interfaces and ties them together for a complete package loading solution.

This approach allows for additional capabilities in the future, such as implementations that work well in web browsers.

**Testing Instructions:**
* Ensure that all unit tests pass
* Try the command-line client as described in the README
* Try [SUSHI's new-fpl branch](https://github.com/FHIR/sushi/tree/new-fpl) that integrates this PR's package loader

_NOTE: Much of this PR contains code that was previously reviewed before merging into the `new-fpl` feature branch. Those reviews, however, typically only involved one reviewer -- and some of the initial code never went through a PR. So... review and test accordingly._

**Related Issue:** Implements #30.

### CODE REVIEW

File: README.md
Comment (mint-thompson) [inline README.md]: This is missing the recently-added `optimize` method from the `PackageLoader` interface.

File: package.json
Comment (jafeltra) [inline package.json:1]: Are we planning to bump the version to v2.0.0-beta.1 after this PR gets merged?

File: README.md
Comment (jafeltra) [inline README.md]: Small typo at the start of the first sentence. I think this probably should be "Loads resources from a passed in implementation..."

File: README.md
Comment (jafeltra) [inline README.md]: The last sentence here references "FHIRPackageLoader", but I think all the other places we talk about the library, we say "FHIR Package Loader" (with spaces).

File: README.md
Comment (jafeltra) [inline README.md]: Small typo in the first sentence.

```suggestion
Finds a loaded package by its name and version, and returns the corresponding [PackageInfo](src/package/PackageInfo.ts) or `undefined` if there is not a match. In the case of multiple matches, the info for last package loaded will be returned. This function supports specific versions (e.g. `1.2.3`), `dev`, `current`, and `current$branchname`.
```

File: src/index.ts
Comment (jafeltra) [inline src/index.ts:1]: This is definitely not a required change, but this file is _so close_ to having all the exports in alphabetical order.

File: src/testPackageLoader.ts
Comment (jafeltra) [inline src/testPackageLoader.ts:1]: Sorry if I'm missing something obvious, but what is the purpose of this file?

File: test/db/SQLJSPackageDB.test.ts
Comment (jafeltra) [inline test/db/SQLJSPackageDB.test.ts]: I think this sort is supposed to include `byLoadOrder()` as well, right?

File: test/cache/DiskBasedPackageCache.test.ts
Comment (jafeltra) [inline test/cache/DiskBasedPackageCache.test.ts]: This may be intentional, but there are no tests here that check what happens when a resource can't be found at the given path. Is there a reason for that? It may not be worth adding them at this point, but just curious.

File: test/cache/DiskBasedPackageCache.test.ts
Comment (jafeltra) [inline test/cache/DiskBasedPackageCache.test.ts:19]: The test coverage also reports that the [if case](https://github.com/FHIR/fhir-package-loader/blob/ae8fa42761c47b94d60b27c1083e8c17508a4b6f/src/cache/DiskBasedPackageCache.ts#L33-L35) for when the `targetDirectory` already exists and gets removed is never tested. Is that something we want to test? I think that case is really only ever reached if we're overwriting a `current` version of a package, but it might be useful to test since it is overwriting something from the cache.

File: src/index.ts
Comment (jafeltra) [reply src/index.ts:1]: Upon further review, alphabetical order is not something we maintain in the other `index.ts`, it just looked more obvious in this one because there are more exports, so I think you can safely ignore my comment here.

File: test/current/BuildDotFhirDotOrgClient.test.ts
Comment (jafeltra) [inline test/current/BuildDotFhirDotOrgClient.test.ts]: This describe block should be outside of the `#downloadCurrentBuild` describe block.

File: test/current/BuildDotFhirDotOrgClient.test.ts
Comment (jafeltra) [inline test/current/BuildDotFhirDotOrgClient.test.ts]: Should this test include a check that the date can be found from a `main` branch and a `master` branch? Or is that overkill here since the implementation is just a regex?

File: src/current/BuildDotFhirDotOrgClient.ts
Comment (jafeltra) [inline src/current/BuildDotFhirDotOrgClient.ts:1]: Okay something very sneaky I just learned -- this file and its corresponding test file are being ignored by prettier right now because the `.prettierignore` file just lists `build*` as the path to ignore. We should probably update that to be `build/*` so this file and the test file get prettier-ified.

File: test/current/BuildDotFhirDotOrgClient.test.ts
Comment (jafeltra) [reply test/current/BuildDotFhirDotOrgClient.test.ts]: If we update the `.prettierignore` file (see other comment below), formatting this file after making this change will be much nicer.

File: test/current/BuildDotFhirDotOrgClient.test.ts
Comment (jafeltra) [inline test/current/BuildDotFhirDotOrgClient.test.ts]: I might be misunderstanding how these tests work, but I'm not sure that I see how these differ. When I debug these tests, I see that the manifest isn't defined for either one, so maybe the second test for `sushi-test-manifest-empty` test needs a different mock? Or maybe I'm misunderstanding how they work.

File: test/registry/FHIRRegistryClient.test.ts
Comment (jafeltra) [inline test/registry/FHIRRegistryClient.test.ts]: Probably don't need this comment. Or maybe we need more from this comment. 🤷

File: src/package/index.ts
Comment (jafeltra) [inline src/package/index.ts]: This is just a question, but is it normal to export `VirtualPackage` from here and from `src/virtual/index.ts`?

**Review by jafeltra — COMMENTED** [review]
This looks really great! This was my first time really digging into this rewrite, and this really breaks out all the separations of concerns and nicely brings them all together for the implementation for SUSHI. I left a few small comments, but none of them are big changes of anything.

File: package.json
Comment (cmoesel) [reply package.json:1]: That's my plan. I didn't want to bump the version and then have to make a bunch of fixes after that. So we'll get the PR through and then bump the version.

File: src/index.ts
Comment (cmoesel) [reply src/index.ts:1]: OK. But in related news... the _imports_ have been bothering me lately. I was thinking that maybe after all of this, I might [configure import ordering in eslint](https://gist.github.com/phatnguyenuit/149094cb3a28e30f5f4c891d264bf7e6) in this project and others. It should probably be a separate PR for most of our projects -- but since this one already changes pretty much every file, maybe I should just put it in as part of this PR. What do you think?

File: test/db/SQLJSPackageDB.test.ts
Comment (cmoesel) [reply test/db/SQLJSPackageDB.test.ts:671]: It was how I manually tested FPL in early versions (before I had a bunch of unit tests and before I updated the CLI). I think it can go away now, so I'll delete it.

File: src/testPackageLoader.ts
Comment (cmoesel) [reply src/testPackageLoader.ts:1]: It was how I manually tested FPL in early versions (before I had a bunch of unit tests and before I updated the CLI). I think it can go away now, so I'll delete it.

File: test/cache/DiskBasedPackageCache.test.ts
Comment (cmoesel) [reply test/cache/DiskBasedPackageCache.test.ts]: In theory, this would only happen if someone deleted the file after it was loaded but before it was queried. That said, I implemented specific behavior for this case, so we should test it.

File: src/index.ts
Comment (jafeltra) [reply src/index.ts:1]: I think I'd be a big fan of that. I like that it sorts the different categories of imports and then sorts alphabetically.

And I guess sorting exports isn't something a lot of people try to do because I don't see an ESLint rule for that.

I don't think I have a preference for if that change happens in this PR or a separate PR. It probably isn't worth holding this PR up just for that, but if you have time while the review is happening, I certainly won't stop you. But I'll also gladly review that in another PR.

File: src/current/BuildDotFhirDotOrgClient.ts
Comment (jafeltra) [reply src/current/BuildDotFhirDotOrgClient.ts:1]: This is probably also getting ignored by ESLint because the config for that lists `build*/*`.

File: test/current/BuildDotFhirDotOrgClient.test.ts
Comment (cmoesel) [reply test/current/BuildDotFhirDotOrgClient.test.ts]: You're right. The network mock needs to to be updated to return null data when that specific manifest was requested.

File: src/package/index.ts
Comment (cmoesel) [reply src/package/index.ts]: No, that's not normal. I think this is a remnant of a refactoring. I originally had `VirtualPackage.ts` in the `src/package` folder -- so it was exporting it then. When I moved `VirtualPackage.ts` to the `src/virtual` folder, the refactoring feature must have rewritten that export and I did not notice it. We don't need it since we are exporting from `src/virtual`.

**Comment by cmoesel** [comment]
Thank you for the very thorough review, @jafeltra. I believe I've addressed all of your findings in a538efd.

**Review by mint-thompson — CHANGES_REQUESTED** [review]
It's finally here! The big FPL refactor! @jafeltra already caught most everything I noticed, but there was one minor discrepancy in the readme that I want to point out.

File: README.md
Comment (cmoesel) [reply README.md]: I love this comment -- because I already fixed it in a538efd! We must have crossed paths! Thanks, @mint-thompson!

**Review by jafeltra — DISMISSED** [review]
Looks great! Thanks for making all those changes! 💯

**Review by mint-thompson — APPROVED** [review]
Hooray for new FPL!

**Review by jafeltra — APPROVED** [review]
Reapproving for such nicely sorted imports! 💯 💯

**Pull request merged**
